### PR TITLE
feat(indexedDB): clean up searchIndex, indexeddb

### PR DESCRIPTION
**Pull request by aewing**
**What this PR does** 📖
- Introduces the `libraries/SatelliteDB` library in place of the thirdparty `dexie` plugin (will revisit plugins as time allows)
- Refactors the SearchIndex class a bit, leveraging `minisearch` in place of the previous `dexie`
- Syncs new conversations, messages, and friends to indexeddb and search indexes

**Which issue(s) this PR fixes** 🔨
[AP-979](https://satellite-im.atlassian.net/browse/AP-979)

**Special notes for reviewers** 🗒️

**Additional comments** 🎤
Feel free to reach out directly with any questions to try to keep the issue clean for other reviewers.

### CODE REVIEW

File: libraries/SatelliteDB/SatelliteDB.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      name: 'foo bar',
+    }
+    await db.upsert('friends', where, update)
+    expect(await db.friends.get('1')).toEqual({
+      ...where,
+      ...original,
+      ...update,
+    })
+  })
+})
```
Comment (stavares843) [inline libraries/SatelliteDB/SatelliteDB.test.ts:70]: thanks for adding the test 🔨

File: libraries/SatelliteDB/SatelliteDB.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+export type KeyValue = {
+  key: string
+  value: string
+  namespace: string
+}
+
+export class SatelliteDB extends Dexie {
+  public isInitialized: boolean
```
Comment (molimauro) [inline libraries/SatelliteDB/SatelliteDB.ts:28]: is this ever set?

Comment (aewing) [reply libraries/SatelliteDB/SatelliteDB.ts]: Nice catch, it is now.

**Review by stavares843 — CHANGES_REQUESTED** [review]
Captura de ecrã 2022-03-21, às 19 04 42

can you clarify exactly what is this PR adding related to the search?

I send messages saying ''a'', searched ''a'', no results found

is the search expected to be working on this PR?

**Review by josephmcg — APPROVED** [review]
I'll let Drew confirm, but I would guess it only searches after a certain character threshold. Searching for a single letter like 'a' wouldn't be very helpful since it would return any word with an 'a'. This seems like an under-the-hood change, UI updates coming in a later PR

**Comment by stavares843** [comment]
well, last time this was tested the search was far from perfect

also, that info is nowhere to be found on the PR, if it has or not a threshold which doesn't even make sense having that because if a user wants to search only by a number or an emoji, he couldn't due to the threshold

**Comment by stavares843** [comment]
also, both locally and CI the imported chats gets broken

Captura de ecrã 2022-03-22, às 04 08 11

is this something expected within this PR?


doesn't seem to be any useful error on the log

**Comment by stavares843** [comment]
i also have that same issue with dexie on specific accounts, more details [here](https://satellite-im.atlassian.net/browse/AP-1105)

**Comment by aewing** [comment]
> also, both locally and CI the imported chats gets broken
> 
> Captura de ecrã 2022-03-22, às 04 08 11
> 
> is this something expected within this PR?
> 
> doesn't seem to be any useful error on the log

my small account imports without issue, so 

> also, both locally and CI the imported chats gets broken
> 
> Captura de ecrã 2022-03-22, às 04 08 11
> 
> is this something expected within this PR?
> 
> doesn't seem to be any useful error on the log

I'm not seeing this issue when importing my account. When I attempt to import the seed phrase provided, I see a series of "Bad API key" messages locally and in the dev environment.

**Comment by stavares843** [comment]
very strange

thanks for checking 🔨

**Comment by aewing** [comment]
> Captura de ecrã 2022-03-22, às 01 27 48
> 
> tested again and ''a'' worked
> 
> but then tested ''test'' and crashed
> 
> Captura de ecrã 2022-03-22, às 01 28 05
> 
> then sent ''SATELLITE'' and searched again by ''SATELLITE'' and crashed
> 
> so this crash's everything you send something > search for that > crash
> 
> reload, crash is gone, send other keyword > search for that one > crash's again

Thanks for finding this. I was able to reproduce this issue and have added a fix (appended the conversation address to new messages before adding them to the index, so the user can be attached in the UI)

File: libraries/SatelliteDB/SatelliteDB.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+   */
+  initializeSchema() {
+    if (this.isInitialized) return
+    this.version(1).stores({
+      conversations: 'key, lastInbound',
+      conversationMessages:
+        '&id, conversation, from, to, at, readAt, type, payload',
+    })
+
+    this.version(2).stores({
```
Comment (WanderingHogan) [inline libraries/SatelliteDB/SatelliteDB.ts:61]: Why not just do this all in one version, eg:
```
    this.version(1).stores({
      conversations: 'key, lastInbound',
      conversationMessages:
        '&id, conversation, from, to, at, readAt, type, payload',
        friends: '&address, textilePubkey, name, photoHash, lastUpdate',
        keyValue: '&key, value, namespace',
    })
```

Backwards compatibility so when this goes in people don't have to wipe their local db?

Comment (WanderingHogan) [reply libraries/SatelliteDB/SatelliteDB.ts:61]: it doesn't really matter for now, will reduce conflicts if we dont combine them right now

**Review by WanderingHogan — APPROVED** [review]

**Comment by WanderingHogan** [comment]
@stavares843 dont worry about anything in search, we have an upcoming task for Joe to refactor/work that

**Comment by stavares843** [comment]
dev

https://user-images.githubusercontent.com/29093946/159589267-af297420-cefb-41c6-af89-dc94a6263977.mov

this branch


https://user-images.githubusercontent.com/29093946/159589365-ca44fb58-6df0-427c-ad1b-b446fc44d283.mov

on dev, you send a glyph - that appears on the front of you
on this branch you send a glyph - that doesn't appear in front of you

**Comment by WanderingHogan** [comment]
i can't for the life of me replicate the problem you are having

**Comment by stavares843** [comment]
so shady

**Comment by stavares843** [comment]
https://user-images.githubusercontent.com/29093946/159595057-3d215310-a1de-4001-b8d2-540727d2791c.mov

flow
- go to dev
- send something
- appears in front of you (A)
- scroll up
- send something - will not appear in front of you, we don't have that type of scroll added yet (B)
- scroll down
- send something
- appears in front of you (because you were all scrolled down like the A above)


then
- go to dev
- send something
- appears in front of you (A)
- scroll up
- send something - will not appear in front of you, we don't have that type of scroll added yet (B)
- scroll down
- send something
- It DOESN'T appears in front of you (and you were all scrolled down like the A above so should be in front of you like dev)

**Comment by aewing** [comment]
> Gravacao.do.ecra.2022-03-22.as.23.45.27.mov
> 
> flow
> 
>     * go to dev
> 
>     * send something
> 
>     * appears in front of you (A)
> 
>     * scroll up
> 
>     * send something - will not appear in front of you, we don't have that type of scroll added yet (B)
> 
>     * scroll down
> 
>     * send something
> 
>     * appears in front of you (because you were all scrolled down like the A above)
> 
> 
> then
> 
>     * go to this branch
> 
>     * send something
> 
>     * appears in front of you (A)
> 
>     * scroll up
> 
>     * send something - will not appear in front of you, we don't have that type of scroll added yet (B)
> 
>     * scroll down
> 
>     * send something
> 
>     * It DOESN'T appears in front of you (and you were all scrolled down like the A above so should be in front of you like dev)

just tested this on https://github.com/Satellite-im/Core-PWA/pull/2407 and a few other branches and the issue is consistent across the open PRs right now.

**Comment by stavares843** [comment]
on dev, is always good for me

Phil is checking as well

**Comment by aewing** [comment]
https://user-images.githubusercontent.com/5033161/159603906-8d06615d-586c-4e79-9de9-6b11677b5e58.mp4

on dev.satellite.one

**Pull request merged**